Simplenote/src/main/java/com/automattic/simplenote/utils/TagUtils.java

```diff
@@ -0,0 +1,70 @@
+package com.automattic.simplenote.utils;
+
+import android.util.Log;
+
+import com.automattic.simplenote.models.Tag;
+import com.simperium.client.Bucket;
+import com.simperium.client.BucketObjectMissingException;
+import com.simperium.client.BucketObjectNameInvalid;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.text.Normalizer;
+import java.util.Locale;
+
+public class TagUtils {
+    /**
+     * Create a tag with the @param key and @param name in the @param bucket.
+     *
+     * @param bucket    {@link Bucket<Tag>} in which to create the tag.
+     * @param name      {@link String} to use as tag name.
+     * @param index     {@link int} to use as tag index.
+     */
+    public static void createTag(Bucket<Tag> bucket, String name, int index) throws BucketObjectNameInvalid {
+        try {
+            Tag tag = bucket.newObject(hashTag(name));
+            tag.setName(name);
+            tag.setIndex(index);
+            tag.save();
+        } catch (BucketObjectNameInvalid e) {
+            Log.e("createTag", "Could not create tag " + "\"" + name + "\"", e);
+            throw new BucketObjectNameInvalid(name);
+        }
+    }
+
+    /**
+     * Create a tag the @param key and @param name in the @param bucket if it does not exist.
+     *
+     * @param bucket    {@link Bucket<Tag>} in which to create the tag.
+     * @param name      {@link String} to use creating the tag.
+     */
+    public static void createTagIfMissing(Bucket<Tag> bucket, String name) throws BucketObjectNameInvalid {
+        try {
+            bucket.getObject(hashTag(name));
+            Log.d("createTagIfMissing", "Tag " + "\"" + name + "\"" + " already exists");
+        } catch (BucketObjectMissingException e) {
+            createTag(bucket, name, bucket.count());
+            Log.d("createTagIfMissing", "Tag " + "\"" + name + "\"" + " does not exist");
+        }
+    }
+
+    /**
+     * Hash the tag @param name with normalizing, lowercasing, and encoding.
+     *
+     * @param name      {@link String} to hash as the tag kay.
+     *
+     * @return          {@link String} hashed to use as tag key.
+     */
+    public static String hashTag(String name) {
+        try {
+            String normalized = Normalizer.normalize(name, Normalizer.Form.NFC);
+            String lowercased = normalized.toLowerCase(Locale.US);
+            String encoded = URLEncoder.encode(lowercased, StandardCharsets.UTF_8.toString());
+            return encoded.replace("*", "%2A").replace("+", "%20");
```

Comment:
this is where we could catch tag names that encode too long…

```java
public static String hashTag(String name) throws TagNameTooLongException {
	…
	String encoded = URLEncoder
		.encode(lowercased, StandardCharsets.UTF_8.toString())
		.replace("*", "%2A")
		.replace("+", "%20");

	// or whatever our length limit is
	// note that our ASCII-encoded value will contain a 1:1 mapping
	// between characters and UTF-16 code units, as reported by `length()`
	if ( encoded.length() > 256 ) {
		throw new TagNameTooLongException();
	}

	return encoded;
```

Reply:
That sounds like a good idea to add a check for encoded length.  If we add it to the `hashTag` method, then we'll need to add a `try`/`catch` block to everywhere `hashTag` is used, which is currently twice in the `Tag` class and twice in the `TagUtils` class.  That's not necessarily a big deal, but might make things confusing when we add some kind of error message to the interface notifying the user that the tag length is too long.

What if we add another method like `hashTagValid` to the `TagUtils` class that does the same thing `hashTag` does, but it returns a boolean if the length is under 256?  We could add that check to the **_Save_** button on the **_Rename Tag_** dialog and the tag creation detection in the **_Add tag..._** field of the editor.  I think we could display an error more gracefully in those cases.

Reply:
This is what I was thinking for the interface.

Rename Tag Dialog| Add Tag Editor
-|-
<a href="https://user-images.githubusercontent.com/3827611/87092289-680db380-c1f8-11ea-96c1-e7db0fd22af5.gif"><img src="https://user-images.githubusercontent.com/3827611/87092289-680db380-c1f8-11ea-96c1-e7db0fd22af5.gif" width="300"></a>|<a href="https://user-images.githubusercontent.com/3827611/87092297-6a700d80-c1f8-11ea-9890-e45a77d46e46.gif"><img src="https://user-images.githubusercontent.com/3827611/87092297-6a700d80-c1f8-11ea-9890-e45a77d46e46.gif" width="300"></a>

Reply:
> What if we add another method like hashTagValid to the TagUtils class that does the same thing hashTag does, but it returns true if the length is less than or equal to 256 and false otherwise?

Seems reasonable to me. It's just a tradeoff between having the compiler catch the bug vs. relying on us to manually do it.

Suggesting that someone contact support makes me nervous, like it could trigger unwanted and unnecessary support volume. I don't know better what to do other than eliminate the constraint the `tagHash` imposes.

Could we be a little fun with it in an informative way?

> Tag length too long (hard to explain); please choose a shorter name or choose something else.

Reply:
In most cases, I wouldn't suggest contacting support immediately.  Also, it's not likely users will run into the issue where they are using a few "characters" and the encoded length is too long.  I'm a little tentative using humor in situations that could be very frustrating for users.  In the example above, it looks like four "characters" to the user and the tag is too long.  They may think you can only have tags with four characters, which is ludicrous.  That may exacerbate the frustration of an error.  That's why I usually go with a dry, straightforward error message.  What about something like this?

> The tag length is too long.  Please, try again with a shorter name.  If the problem continues, contact us at support@simplenote.com for help.